include/frozen/bits/basic_types.h

```diff
@@ -103,46 +105,43 @@ class carray {
   using const_pointer = const value_type *;
   using iterator = pointer;
   using const_iterator = const_pointer;
-  using reverse_iterator = std::reverse_iterator<iterator>;
-  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
   using size_type = std::size_t;
   using difference_type = std::ptrdiff_t;
 
   // Constructors
-  constexpr carray(void) = default;
-  template <std::size_t M>
-  constexpr carray(T const (&init)[M])
+  constexpr carray() = default;
+  constexpr carray(const value_type& val)
+    : carray(val, std::make_index_sequence<N>()) {}
+  template <typename U, std::enable_if_t<std::is_convertible<U, T>::value, std::size_t> M>
+  constexpr carray(U const (&init)[M])
     : carray(init, std::make_index_sequence<N>())
   {
     static_assert(M >= N, "Cannot initialize a carray with an smaller array");
   }
-  template <std::size_t M>
-  constexpr carray(std::array<T, M> const &init)
-    : carray(&init[0], std::make_index_sequence<N>())
+  template <typename U, std::enable_if_t<std::is_convertible<U, T>::value, std::size_t> M>
+  constexpr carray(std::array<U, M> const &init)
+    : carray(init.begin(), std::make_index_sequence<N>())
   {
     static_assert(M >= N, "Cannot initialize a carray with an smaller array");
   }
-  constexpr carray(std::initializer_list<T> init)
+  template <typename U, std::enable_if_t<std::is_convertible<U, T>::value>* = nullptr>
+  constexpr carray(std::initializer_list<U> init)
     : carray(init.begin(), std::make_index_sequence<N>())
   {
     // clang & gcc doesn't recognize init.size() as a constexpr
     // static_assert(init.size() >= N, "Cannot initialize a carray with an smaller initializer list");
   }
+  template <typename U, std::enable_if_t<std::is_convertible<U, T>::value>* = nullptr>
+  constexpr carray(const carray<U, N>& rhs)
+    : carray(rhs.begin(), std::make_index_sequence<N>())
+  {
+  }
 
   // Iterators
   constexpr iterator begin() noexcept { return data_; }
   constexpr const_iterator begin() const noexcept { return data_; }
-  constexpr const_iterator cbegin() const noexcept { return data_; }
```

Comment:
To be sure: getting rid of those because  `carray` is not part of the official API and these methods are unused?

Reply:
It not being part of the public/official API is why it's possible to get rid of these. The reason for actually doing so is because doing it in this location doesn't really save any effort nor prevent any duplication in the container frontend types (it's just different duplication). While at the same time, when adding an extra storage-container type, this is code that needs to be duplicated and thus is extra code to hide bugs in.